### DISCUSSION TITLE
db.obj.create() Works in Development But Not When Deployed

### DISCUSSION THREAD

**Discussion by lynixiox** [prisma/prisma#23361] [Q&A] [answered]
### Question

I am trying to create a voting object in the database. The object creates just fine in development but not when deployed. I have some logs . In development it logs before and after the completion. In deployment it logs only before.It does catch and logs the following in deployment 
```
PrismaClientInitializationError: Invalid prisma.vote.create() invocation:
Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)
    at _n.handleRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:123:7154)
    at _n.handleAndLogRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:123:6188)
    at _n.request (/var/task/node_modules/@prisma/client/runtime/library.js:123:5896)
    at async l (/var/task/node_modules/@prisma/client/runtime/library.js:128:10871)
    at async a (/var/task/.next/server/app/competition/contestant/[id]/page.js:1:12755) {
  clientVersion: '5.10.2',
  errorCode: undefined
}
```

There are no hoops or any different configs from my development to my deployment and I am able to successfully create other objects in the deployment. Am I missing a config somewhere?

### Information about Prisma Schema, Client Queries and Environment 

Prisma schema
```prisma
model Bundle {
  id String @id @default(cuid())
  bundleId String
  totalValue String
  pricePerVote String
  competition Competition @relation(fields: [bundleId], references: [id], onDelete: Cascade)
  vote Vote[] 
  @@index([bundleId])
  @@unique([id])
}

model Contestant{
  id String @id @default(cuid())
  contestantId String 
  name String
  age String
  votes Vote[]
  competition Competition @relation(fields: [contestantId], references: [id], onDelete: Cascade)
  @@index([contestantId])
  @@unique([id])
}

model Vote{
  id String @id @default(cuid())
  voteId String
  contestantId String
  voterName String
  amountOFVotes Int
  bundle Bundle @relation(fields: [voteId], references: [id], onDelete: Cascade) 
  contestant Contestant @relation(fields: [contestantId], references: [id], onDelete: Cascade)
  @@index([contestantId])
  @@index([voteId])
  @@unique([id])
}

```

Action vote.ts
```ts
import { ContestantDef } from "@/lib/definitions"
import { Bundle } from "@prisma/client"

export const createVote = async({
    amountOFVotes, 
    voterName, 
    bundle, 
    contestant }:{
        amountOFVotes: number
        voterName: string
        bundle: Bundle,
        contestant: ContestantDef
    }) => {
    try{
        console.log("creating vote, ", amountOFVotes)

        await db.vote.create({
            data: {
                amountOFVotes,
                voterName,
                bundle: {
                    connect: {id: bundle.id}
                },
                contestant:{
                    connect: {id: contestant!.id}
                }
            }
        })

        console.log("Finished Creating Vote")

    }catch(error){
        console.log(error)
    }

}
```
ContestantDef
```ts
export type ContestantDef = Prisma.PromiseReturnType<typeof getContestantByid>  
```

package.json
```js
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "postinstall": "prisma generate"
  },
```

Payload
```ts
"use server"

import { ContestantDef } from "@/lib/definitions"
import { createVote } from "./vote"
import { Bundle } from "@prisma/client"

export const verify = async({amountOFVotes, voterName, bundle, contestant}:
  {
    amountOFVotes: number,
    voterName : string,
    bundle: Bundle,
    contestant:ContestantDef
  }) => {
  createVote({amountOFVotes: amountOFVotes,
    voterName: voterName,
    bundle: bundle!,
    contestant: contestant
  })
}
```
- OS: Manjaro
- Database: PlaneScale
- Node.js version: v-20.10.0
- Prisma version: 5.10.2

**Comment by ludralph** [comment]
Hi @lynixiox 👋 

Thank you for raising this question.

Can you consider increasing your[ connection limit](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#optimizing-the-connection-pool) to handle potential spikes in concurrent votes. The current timeout is 10 seconds. Experimenting with increasing the [timeout](https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections#increasing-the-pool-timeout) value (within reasonable limits) might help if the database takes slightly longer to respond in the deployment environment.

**Reply by lynixiox** [reply]
Hi @ludralph :wave: 

Thank you for getting back to me.

I have increased the connection limit to 15 and the pool_timeout to 40 to see if will work. Its been 20 minutes and it has not even thrown the error. it has only logged `creating vote`  and is now hanging :thinking: do you think my timeout is to high now?

**Reply by janpio** [reply]
What do you mean by "hanging" - the error message again?

The error message only happens if there are more than 5 (or 15 after your config change) connections open at the same time running a query already.

**Reply by lynixiox** [reply]
Hi @janpio 

Apologies. What I mean by hanging is that the `error` is never logged. It only logs `creating vote` and then doesn't do anything else. The program continues to work just fine this `action` essentially remains unresolved

**Reply by janpio** [reply]
After your `pool_timeout` it should still turn into an error if that is the root of your problem. 20 minutes is definitely not correct and is very unexpected. I suggest you decrease the pool_timeout again - waiting 40 seconds for a page will not be good DX anyway.

You should try to find a solution to the problem that cause the connection pool timeout error message in the first place. (See my explanation what causes this above)

**Reply by lynixiox** [reply]
I have removed it entirely and have tried a suggestion from #19723 around 30 minutes ago. `&ssl=true` however this results in the same behavior where no `error` has been logged. I have also tried going back to the original DATABASE_URL that logged the initial `error` however nothing again was logged.  I am able to Create Competitions and Users just fine it is just the Vote that will not execute on deployment

**Reply by janpio** [reply]
That sounds super weird, and honestly like something we can help you debug much.

Some pointers on what I would try:
Does the request succeed when you comment out the Prisma Client query?
What about when you replace it with another model based one?
Or when you change the query slightly?

**Reply by lynixiox** [reply]
I have updated the try catch as follows 
```ts
    try{
        console.log("Looking for Votes, ", amountOFVotes)

        await db.vote.findMany({
            where: {contestant: {id: contestant!.id}}
        })

        console.log("found the votes")

    }catch(error){
        console.log(error)
    }
```
it has been 4 minutes and it logged `Looking for votes` and nothing else since. When I ran the same request on `localhost` it loged `found the votes` almost instantly

the following request works perfectly on deployment in another part of the program
```ts
    const contestantData = await db.contestant.findMany(
      {
        where: {
          competition: {id}
        },
        include: {votes: true}
      }
    );
    console.log(contestantData)
    return contestantData
```
I am going to add the above call to the function giving me an issue and will let you know if it logs anything. It has been another 4 minutes since I finished typing this and still no error log on deployment. I have removed all `limits` and `pool_timeouts`. I can also confirm that the .env files are identical at this point in time

**Reply by lynixiox** [reply]
I updated the problem function with the `request` that we know works when `Deployed`
```ts
        console.log("Looking for Contestants ", amountOFVotes)

        const contestantData = await db.contestant.findMany(
            {
              where: {
                competition: {id: contestant?.id}
              },
              include: {votes: true}
            }
          );
          console.log(contestantData)

        console.log("found the Contestants")
```

Same result as before. `localhost` logs the data just fine `Deployment` logs `Looking for Contestants` and then nothing else

**Reply by janpio** [reply]
So the same query that works in one part of the app does not work in _this_ part of the app?
That does not really make sense, correct?
Is anything else happening in this function before? Some "middleware" that could interfere?

**Reply by lynixiox** [reply]
Yes this whole issue is super weird and doesn't make sense. You are correct in saying the query works in one part of the app and not in this part. 

There is not any `middleware` for this you pretty much have everything for this query on here except the initial call which looks like this 

```ts
    const onSuccess = () => {
     verify({amountOFVotes: numberOfBundles*amount , voterName: name, bundle: bundle!, contestant: contestant}  )
    };
```

this is called after the payment gateway `PayStack` succeeds in its transaction.  I'm going to explore the possibility that `vercel` might be the culprit here and spin up an `EC2` instance to see if I see a similar result to this issue. Let me know if you have any other ideas or if theres anything else I can do for you to help with this issue.

**Reply by lynixiox** [reply]
Deployed the working query to `AWS Amplify` 

```ts
        console.log("Looking for Contestants ", amountOFVotes)

        const contestantData = await db.contestant.findMany(
            {
              where: {
                competition: {id: contestant?.id}
              },
              include: {votes: true}
            }
          );
          console.log(contestantData)

        console.log("found the Contestants")
```
At first it was the same result, logs `Looking for Contestants` then nothing. However when I refreshed the page the error log popped up 

```
Invalid prisma.contestant.findMany() invocation:
Can't reach database server at `aws.connect.psdb.cloud`:3306
```
 Now the confusing part for me is that in order to get the point where the function `createVote()` is called I first go past a page that successfully runs `prisma.contestant.findMany()` everytime. However when the `createVote()` function is called it fails every time. Is this looking like a `Prisma` , `Nextjs` or `Pscale` problem :thinking: ?

**Comment by lynixiox** [comment, marked as answer]
Added `"use server"` to the `action.ts` file. Removed the hop from `OnSuccess` to `verify.ts` and call `createVote()` directly `OnSuccess` 

Extremely weird issue not sure if anything else needs to be discussed futher
